pkg/operator/apiserver/controller/workload/workload.go

```diff
@@ -0,0 +1,447 @@
+type SyncFunc func() (*appsv1.Deployment, bool, []error)
+
+// Controller is a generic workload controller that deals with Deployment resource.
+// Callers must provide a sync function for delegation. It should bring the desired workload into operation.
+// The returned state along with errors will be converted into conditions and persisted in the status field.
+type Controller struct {
+	name                 string
+	operatorNamespace    string
+	targetNamespace      string
+	targetOperandVersion string
+	// operandNamePrefix is used to set the version for an operand via versionRecorder.SetVersion method
+	operandNamePrefix string
+
+	operatorClient               v1helpers.OperatorClient
+	kubeClient                   kubernetes.Interface
+	openshiftClusterConfigClient openshiftconfigclientv1.ClusterOperatorInterface
+
+	syncFn             SyncFunc
+	queue              workqueue.RateLimitingInterface
+	eventRecorder      events.Recorder
+	versionRecorder    status.VersionGetter
+	preRunCachesSynced []cache.InformerSynced
+}
+
+// NewController creates a brand new Controller instance.
+//
+// the "name" param will be used to set conditions in the status field. It will be suffixed with "WorkloadController",
+// so it can end up in the condition in the form of "OAuthAPIWorkloadControllerDeploymentAvailable"
+//
+// the "operatorNamespace" is used to set "version-mapping" in the correct namespace
+//
+// the "targetNamespace" represent the namespace for the managed resource (DaemonSet)
+func NewController(name, operatorNamespace, targetNamespace, targetOperandVersion, operandNamePrefix string,
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	syncFn SyncFunc,
+	openshiftClusterConfigClient openshiftconfigclientv1.ClusterOperatorInterface,
+	eventRecorder events.Recorder,
+	versionRecorder status.VersionGetter) *Controller {
+	controllerRef := &Controller{
+		operatorNamespace:            operatorNamespace,
+		name:                         fmt.Sprintf("%sWorkloadController", name),
+		targetNamespace:              targetNamespace,
+		targetOperandVersion:         targetOperandVersion,
+		operandNamePrefix:            operandNamePrefix,
+		operatorClient:               operatorClient,
+		kubeClient:                   kubeClient,
+		syncFn:                       syncFn,
+		openshiftClusterConfigClient: openshiftClusterConfigClient,
+		eventRecorder:                eventRecorder.WithComponentSuffix("workload-controller"),
+		versionRecorder:              versionRecorder,
+		queue:                        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), name),
+	}
+
+	return controllerRef
+}
+
+func (c *Controller) sync() error {
+	operatorSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+
+	if run, err := c.shouldSync(operatorSpec); !run {
+		return err
+	}
+
+	if fulfilled, err := c.preconditionFulfilled(operatorSpec); !fulfilled {
```

Comment:
I think I am going to add a custom function for `precoditions`, for example when a config hasn't been populated there is no need to run the `sync` function. That function would be provided by the callers.